### ISSUE TITLE
[FNA] [HARD] SpriteBatch/SpriteFont Rewrite

### ISSUE THREAD

**Issue by flibitijibibo** [flibitijibibo/FNA#246]
Difficulty: This task is considered a HARD task. If MEDIUM tasks separate the men from the boys, HARD tasks separate Dark Souls fans from actual serial killers that wear proper armor for reasons yet to be explained.

I've been working on a new implementation of the SpriteBatch/SpriteFont classes:

https://github.com/flibitijibibo/SpriteBatchRewrite

These two classes are probably among the oldest classes in the MonoGame source tree right now, and their age has started to show quite a bit... on top of numerous accuracy issues (especially for SpriteFont), there are also a lot of little details that needed fixing. But, rather than edit the current files, I just wrote some new ones from scratch.

Here's what's different:
- Use of VertexBuffer/IndexBuffer, rather than client-side arrays. This is more accurate and is much nicer for the driver to work with.
- Much of the functionality currently comes from the work found in these notes:

https://github.com/flibitijibibo/SpriteBatchRewrite/blob/master/SpriteBatch.cs#L20

It should be expected that accuracy overall is _significantly_ improved. However, we've got some things left to do before we can use this in FNA. Primarily, performance needs to be improved dramatically. Right now the rewrite's Immediate drawing is 25% slower, and the Deferred drawing is almost 100% (!!!) slower, despite the relatively optimized use of the buffer objects. I don't expect perf to be as good as XNA4, due to use of pure managed code in addition to the OpenGLDevice's own perf issues (which is why I wanted to write this in the first place), but we need to keep perf as good as we found it if we want to use this.

apitrace/vogl output suggests that we're doing really well with the GL; the issue is, similar to the OpenGLDevice, a CPU-bound performance hit. There's a chance that this is the result of us doing things more accurately, but I highly doubt that it could warrant such a huge performance drop.

SpriteBatch.PushSprite and SpriteBatch.RenderBatch are the slowest methods in the rewrite by a pretty wide margin (the Array.Sort calls used to sort for Texture/BackToFront/FrontToBack seem pretty slow too). There are definitely lots of ways to optimize this, so look to those two blocks of code first.

It should be possible to drop these two files into an upstream FNA revision and build without changing too many files (SpriteFontReader might need one line fixed for textureValue), so feel free to work on the rewrite that way. You may want to keep the old files handy so you can run performance comparisons.
